test/topology_custom/test_mv_backlog.py

```diff
@@ -0,0 +1,112 @@
+                    # doesn't have "continue 2" so let's just set val to 0...
+                    val = 0
+                    break
+        total += float(val)
+    return total
+
+def get_replicas(cql, key):
+    return cql.cluster.metadata.get_replicas("ks", Int32Type.serialize(key, cql.cluster.protocol_version))
+
+async def enable_on_servers(manager, servers, injection, oneshot=False):
+    errs = [manager.api.enable_injection(s.ip_addr, injection, oneshot) for s in servers]
+    return await asyncio.gather(*errs)
+async def disable_on_servers(manager, servers, injection):
+    errs = [manager.api.disable_injection(s.ip_addr, injection) for s in servers]
+    return await asyncio.gather(*errs)
+
+def get_remote_key(cql, local_node):
+    i = 0
+    while local_node == get_replicas(cql, i)[0]:
+        i = i + 1
+    return i
+
+# This test reproduces issue #18542
+# In the test, we create a table and perform a write to it a couple of times
+# Each time, we check that a view update backlog on some shard increased
+# due to the write.
+@pytest.mark.asyncio
+@skip_mode('release', "error injections aren't enabled in release mode")
+async def test_view_backlog_increased_after_write(manager: ManagerClient) -> None:
+    node_count = 2
+    # Use a higher smp to make it more likely that the writes go to a different shard than the coordinator.
+    servers = await manager.servers_add(node_count, cmdline=['--smp', '5'])
+    cql = manager.get_cql()
+    await cql.run_async(f"CREATE KEYSPACE ks WITH replication = {{'class': 'SimpleStrategy', 'replication_factor': 1}}")
+    await cql.run_async(f"CREATE TABLE ks.tab (key int, c int, v text, PRIMARY KEY (key, c))")
+    await cql.run_async(f"CREATE MATERIALIZED VIEW ks.mv_cf_view AS SELECT * FROM ks.tab "
+                    "WHERE c IS NOT NULL and key IS NOT NULL PRIMARY KEY (c, key) ")
+    await wait_for_view(cql, 'mv_cf_view', node_count)
+
+    key = int(time.time())
+    logger.info(f"Base table key: {key}")
+    local_node = get_replicas(cql, key)[0]
+    # Only remote updates hold on to memory, so make the update remote
+    remote_key = get_remote_key(cql, local_node)
+    for v in [1000, 4000, 16000, 64000, 256000]:
+        # Make sure the view update backlog is still high (view udpate hasn't finished) when the replica returns
+        await enable_on_servers(manager, servers, "never_finish_remote_view_updates")
+
+        await cql.run_async(f"INSERT INTO ks.tab (key, c, v) VALUES ({key}, {remote_key}, '{v*'a'}')")
+        # The view update backlog should increase on the node generating view updates
+        view_backlog = get_metric(local_node.address, 'scylla_storage_proxy_replica_view_update_backlog')
+        # The read view_backlog might still contain backlogs from the previous iterations, so we only assert that it is large enough
+        assert view_backlog > v
+        await disable_on_servers(manager, servers, "never_finish_remote_view_updates")
```

Comment:
What's the point of disabling the injection at the end of each iteration if you'll immediately enable it when the next iteration starts? 

As you said in the previous comment, you don't know if you leave enough time for previous write to realize they can finish - you probably don't (the write has that 100ms sleep, and if you don't sleep 100ms here, this sleep might not finish).

Reply:
Temporarily disabling this injection allows the previous write to finish, but as you've noticed, it's still possible that the update doesn't finish before the next write, particularly in debug (though the injection is added after performing the view update so the view update semaphore units should be released rather quickly). I could block all the view updates (there's just a few of them anyway) until the test finishes and check that all of them are still contributing to the backlog - I'll do that in the next update as I don't see any downsides to it and it will remove this unnecessary complication from the test

Reply:
IMO you could:

- Enable the error injection at the beginning of the loop and not disable it in the middle,
- Instead of introducing the new metric, you could fetch the value of `view_update_backlog`. Don't sum the per-shard values, keep them in per-shard form
- Compare the value from the previous iteration with the value of the current iteration. Observe that the backlog only increased on one shard and by more than `v`. If possible, you could also check that the backlog increased on the correct shard on the replica (simpler variant - that it always increases on the same shard).

Reply:
I changed the test so that we don't disable the injection at all. I also changed the test to read the metric from just one shard, but I can't change the test to use the already existing `view_update_backlog` metric, because it's not the value affected by this patch.